runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -352,20 +407,18 @@ public RemoteBundle getBundle(
         // The blocking queue of caches for serving multiple bundles concurrently.
         currentCache = availableCaches.take();
         client = currentCache.getUnchecked(executableStage.getEnvironment());
-        client.ref();
 
         currentClient = preparedClients.get(client);
         if (currentClient == null) {
           // we are using this client for the first time
           preparedClients.put(client, currentClient = prepare(client, executableStage));
           // cleanup any expired clients
-          preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() == 0);
+          preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() <= 0);
```

Comment:
Why the change from `==` to `<=`?

Reply:
I prefer including the entire range whenever I write such checks. Just a way to guard against future changes where we might countdown after the value has reached zero.

Reply:
On the flip side, this could hide a bug in the reference counting. That could be covered with tests though. Not important for this PR.

Reply:
I'll also revert this change to not distract from the main objective of this PR.

Reply:
Not important, can stay as is. What would be good to check is that the environment expiration still works.